pkgs/tools/admin/trivy/default.nix

```
@@ -29,9 +32,22 @@ buildGoModule rec {
     "-X=github.com/aquasecurity/trivy/pkg/version.ver=v${version}"
   ];
 
+  nativeBuildInputs = [ installShellFiles ];
+
   # Tests require network access
   doCheck = false;
 
+  postInstall =
+    let
+      trivy = if stdenvNoCC.buildPlatform.canExecute stdenvNoCC.hostPlatform then "\"$out\"" else buildPackages.trivy;
+    in
+    ''
+      installShellCompletion --cmd ${pname} \
+        --bash <(${trivy}/bin/${pname} completion bash) \
+        --fish <(${trivy}/bin/${pname} completion fish) \
+        --zsh <(${trivy}/bin/${pname} completion zsh)
+    '';
+
```

I guess if you're going that far to not break the build, then perhaps also check that `buildPackages.trivy` can actually be built/isn't broken?

I usually just see people drop the shell completions when cross-compiling.

---

I've copied this approach from #225363, so I thought it's an accepted pattern, in a way. I'm mostly concerned about cross-compiling stuff from x86_64 -> X. I see why you're asking, however, in this specific case of trivy, that check wouldn't be relevant, since AFAIU it wouldn't unblock any host/build platform combination which would not already be working without it. Trivy can probably be compiled on every platform that's supported by Go, and without having Go available on the host platform, cross-compiling wouldn't be possible anyways.

---

I hadn't seen it before.

I have a small nit: don't use `pname` to refer to the executable. Either set `meta.mainProgram` and use `lib.getExe` or write it out explicitly:

```suggestion
  postInstall =
    let
      trivy = if stdenvNoCC.buildPlatform.canExecute stdenvNoCC.hostPlatform then "\"$out\"" else buildPackages.trivy;
    in
    ''
      installShellCompletion --cmd ${pname} \
        --bash <(${trivy}/bin/trivy completion bash) \
        --fish <(${trivy}/bin/trivy completion fish) \
        --zsh <(${trivy}/bin/trivy completion zsh)
    '';

```

---

Will do! Just curious, what's the rationale not to use pname?

---

Because pname can be `open-broadcaster-software` then the executable can be `obs` for example 

As a note meta.mainProgram will default to pname so you only need to set it if there is an exception.

---

> As a note meta.mainProgram will default to pname so you only need to set it if there is an exception.

Not since #246386, this handy behaviour has been deprecated.

---

I was more talking about `meta.mainProgram` itself falling back to `pname` but now I can't find the nix for that

---

It doesn't, `nix run` has the fallback. `lib.getExe` used to match it, but is deprecating that behavior.